### ISSUE TITLE
TypeError: undefined is not an object (evaluating 'this.sheet.insertRule') on mobile safari

### ISSUE THREAD

**Issue by chexee** [threepointone/glamor#58]
This error occurs in only on our (sourcegraph.com) production set up, when glamor is in `speedy` mode. If its relevant, we also use TypeScript.

This can be reproduced in latest Safari and Chrome on iOS 10.0.1 on iPad and iPhone.

It does not appear to be an issue on some other mobile devices, e.g., Nexus 5x / Chrome.

This is the JS error I am getting on my iPad in Safari:

![3f5c9cb0-802c-11e6-8be3-898359706305](https://cloud.githubusercontent.com/assets/285836/18759170/1963310c-80b0-11e6-9d3e-7300591e8eac.png)

For now, we've disabled `speedy` mode, but it is, of course slower.

**Comment by threepointone** [comment]
I'll have a look today, thanks for the catch!

**Comment by threepointone** [comment]
btw I'm so happy to see you're using this in production! do you have any feedback? anything I can make better? can you tell me more about your usage? thank you!

**Comment by chexee** [comment]
I love the syntax, but definitely took a little longer to get a sense of what exactly was happening :)

Feel free to email me and we can chat more outside of the issue :)

**Comment by threepointone** [comment]
so my unit tests are passing in ios10 with speedy true/false. I'll work on making a standalone sample in a while, unless you have a small test case that can reproduce this? 

And I'd love to chat via mail. Will mail to the id on your profile, thanks!

**Comment by chexee** [comment]
I'll see if I can muster up a test case for this today ~

**Comment by threepointone** [comment]
This bug is particularly perplexing because there are only 2 instances of `this.sheet.insertRule` [in the source](https://github.com/threepointone/glamor/blob/master/src/sheet.js). One of them runs only when `typeof document === 'undefined'` ie on the server, and the other is wrapped in a `try/catch`. 

I'll keep digging, but let me know if you can reproduce this bug (if you can deploy your site to a secret url or something so I can step through it, would also work)

**Comment by chexee** [comment]
Okay, so digging a bit more, here's the specific line that causes the error in production mode (obviously after webpack has compiled ... poking around to see if I can find the uncompiled line. will post more soon)
![screenshot 2016-09-26 12 00 51](https://cloud.githubusercontent.com/assets/285836/18847957/0c4a3f7c-83e1-11e6-9923-da8a98d7b716.png)

**Comment by threepointone** [comment]
ok this helps! can you tell me if you're doing anything special while loading the script? async or whatever? is it in an iframe or something? or targeting one?

**Comment by chexee** [comment]
Nope - no iframe or targeting iframe.

I ran it locally without minification and traced the error back to https://github.com/threepointone/glamor/blob/master/src/sheet.js#L118

Here's the log from Safari:
![screenshot 2016-09-26 12 17 33](https://cloud.githubusercontent.com/assets/285836/18848491/835ab59a-83e3-11e6-9419-4e9e41aeb14f.png)

**Comment by threepointone** [comment]
could you share your setup with ngrok or something similar so I could have a look? https://ngrok.com/

**Comment by threepointone** [comment]
it's late here and I'm wiped out, so I'm going to crash for a few hours. I made a small sample repo here https://github.com/threepointone/glamor-58 it seems to work on my iphone 6/ios 10, but please give it a shot and / or try to break it. cheers!

**Comment by shurcooL** [comment]
Not sure if this is obvious or not, but here's an observation. Please ignore if it's unhelpful.

The error message says:

> undefined is not an object (evaluating 'this.sheet.insertRule')

I think it's referring to `this.sheet` being undefined (for some reason). It's not related to `insertRule`.

**Comment by threepointone** [comment]
yup, got that. that implies that `sheetForTag()` is failing/returning null, but I'm not able to reproduce it. 
gave both of you commit rights to the sample repo, lemme know if you can reproduce this. thanks!

**Comment by chexee** [comment]
Hey – i'll bang on this tomorrow and try to repro. Thanks for your time and effort :)

**Comment by threepointone** [comment]
I'm very eager to solve this for you folks, so please feel free to bug me any time.

**Comment by vdanchenkov** [comment]
I'm curious too. Tried some ideas via iOS 10 emulator on browserstack, no luck so far.

**Comment by chexee** [comment]
Sorry! This week has been pretty busy. I probably won't get more time to try to repro until next week. I'll plan to bang on this more on Sunday/Monday and see if I can't track down the issue in your sample repo by then ~

**Comment by threepointone** [comment]
take your time!

**Comment by KyleAMathews** [comment]
Which version of iOS are you on @chexee? I can't reproduce this on a site I'm building with Glamor on the latest iOS — https://loanpal.github.io/frontend/

**Comment by shurcooL** [comment]
It was iOS 10.0.1 or 10.0.2.

I tried the link you posted above, and I can't reproduce it there either.

**Comment by gnoff** [comment]
I am getting bit by this inconsistently on the latest chrome. If I refresh the page i get the insertRule error but if I reload by activating the url bar and hitting enter it works.

**Comment by threepointone** [comment]
@gnoff can you point me to a page where this is happening?

**Comment by everdance** [comment]
Hi @threepointone we did some debug on our site, the error happens within the following code:

I suppose this ** document.styleSheets[i].ownerNode === tag ** is not consistently returns true upon our site initialization
```
function sheetForTag(tag) {
  for (var i = 0; i < document.styleSheets.length; i++) {
    if (document.styleSheets[i].ownerNode === tag) {
      return document.styleSheets[i];
    }
  }
}
```
`sheetForTag(this.tags[0]);` returns **undefined** in `inject` call

```
 _createClass(StyleSheet, [{
    key: 'inject',
    value: function inject() {
      var _this = this;

      if (this.injected) {
        throw new Error('already injected stylesheet!');
      }
      if (isBrowser) {
        // this section is just weird alchemy I found online off many sources 
        this.tags[0] = makeStyleTag();
        // this weirdness brought to you by firefox 
        this.sheet = sheetForTag(this.tags[0]);
      } else {
        // server side 'polyfill'. just enough behavior to be useful.
        this.sheet = {
          cssRules: [],
          insertRule: function insertRule(rule) {
            // enough 'spec compliance' to be able to extract the rules later  
            // in other words, just the cssText field 
            _this.sheet.cssRules.push({ cssText: rule });
          }
        };
      }
      this.injected = true;
```

`insert` throws with **sheet** undefined
```
    key: 'insert',
    value: function insert(rule) {

      if (isBrowser) {
        var _context;

        // this is the ultrafast version, works across browsers 
        if (this.isSpeedy && this.sheet.insertRule) {
          this._insert(rule);
        }
        // 
...
```

**Comment by everdance** [comment]
so I also tested on different browsers, **document.styleSheets[i].ownerNode===tag** only works for firefox, not work on chrome and safari

**Comment by threepointone** [comment]
thanks @everdance , this helps a lot! odd that it wouldn't work consistently. I'll think of a workaround.

**Comment by threepointone** [comment]
I just published a version 2.17.6 with a possible fix, would you folks give it a spin and see if it's fixed the issue? plsthx.

**Comment by everdance** [comment]
sure, will try and update you.

**Comment by everdance** [comment]
@threepointone `2.17.6` is still unavailable on npm

**Comment by threepointone** [comment]
dammit. try again please.

**Comment by everdance** [comment]
cool, works now with `2.17.6`, thanks @threepointone

**Comment by chexee** [comment]
Not sure if helpful, but here's an update from our side – this looks like it resolved itself for us with the latest iOS update: 

I tried to reproduce it locally and couldn't. When I ran into the bug initially, it was 100% reproducible on my iPad mini with iOS 10.0.1. Decided to re-enable speedy mode now on our production site, and not seeing the issue anymore. By now I'm running iPad mini on iOS 10.0.2, also tested on iPhone with iOS 10.0.2, on desktop Safari on macOS Sierra 10.12 (16A323), etc.

I haven't seen it on Chrome, but we'll update to the latest in case it comes up
